### DISCUSSION TITLE
Automated fault tolerance test

### DISCUSSION THREAD

**Discussion by sonallux** [jvalue/open-data-service#297] [Show and tell]
Hey everyone, 

to support and track the progress of my thesis about fault tolerance in microservice-based architectures I have written some [automated fault tolerance tests](https://github.com/sonallux/open-data-service/tree/fault-tolerance-test/fault-tolerance-test) for the ODS. They are running on GitHub actions, so this status badge will always reflect the latest status of my fault tolerance tests:

[![Fault tolerance test](https://github.com/sonallux/open-data-service/workflows/Fault%20tolerance%20test/badge.svg?branch=fault-tolerance-test)](https://github.com/sonallux/open-data-service/actions?query=workflow%3A%22Fault+tolerance+test%22)

### Restart test
Tests that every service can be restarted, without causing any problems. After starting the ODS it performs some requests to add some state. Then one service at a time is stopped and restarted. After the service is running again checks are performed to validate that no data is lost and the ODS is still completely functional. Because the services are currently not replicated, the test ensures that no traffic is performed between stopping and the successful restart of service. Persistent storage is also not erased and therefore no state inconsistencies can arise.

Following bugs have been detected by this test:
- [x] closing the database connection will crash the microservice jvalue/node-dry-pg#2
- [x] Already consumed events are redelivered after a service restart
- [x] Connections to RabbitMQ are not restarted if they brake

### Scheduler initialization test
This test is for issue #284

**Comment by lunedis** [comment]
Really nice work!

If GitHub is running their GitHub Action runners inside e.g. kubernetes, and they use containers to create the jobs, and you use containers inside such a job to run services, how many levels deep are we and when do we reach the limbo? :smile:
